cardano-api/src/Cardano/Api/IPC/Monad.hs

```diff
@@ -74,38 +85,59 @@ setupLocalStateQueryExpr ::
   -> Net.Query.LocalStateQueryClient (BlockInMode mode) ChainPoint (QueryInMode mode) IO ()
-setupLocalStateQueryExpr waitDone mPointVar' resultVar' f =
+setupLocalStateQueryExpr waitDone mPointVar' resultVar' ntcVersion mapQueryError f =
   LocalStateQueryClient . pure . Net.Query.SendMsgAcquire mPointVar' $
     Net.Query.ClientStAcquiring
-    { Net.Query.recvMsgAcquired = runContT (runLocalStateQueryExpr f) $ \result -> do
-        atomically $ putTMVar resultVar' (Right result)
+    { Net.Query.recvMsgAcquired = runContT (runReaderT (runLocalStateQueryExpr f) ntcVersion) $ \result -> do
+        atomically $ putTMVar resultVar' result
         void $ atomically waitDone -- Wait for all protocols to complete before exiting.
         pure $ Net.Query.SendMsgRelease $ pure $ Net.Query.SendMsgDone ()
 
     , Net.Query.recvMsgFailure = \failure -> do
-        atomically $ putTMVar resultVar' (Left failure)
+        atomically $ putTMVar resultVar' (Left (mapQueryError (QueryErrorAcquireFailure failure)))
         void $ atomically waitDone -- Wait for all protocols to complete before exiting.
         pure $ Net.Query.SendMsgDone ()
     }
 
--- | Use 'queryExpr' in a do block to construct monadic local state queries.
-queryExpr :: QueryInMode mode a -> LocalStateQueryExpr block point (QueryInMode mode) r IO a
-queryExpr q =
-  LocalStateQueryExpr . ContT $ \f -> pure $
-    Net.Query.SendMsgQuery q $
-      Net.Query.ClientStQuerying
-      { Net.Query.recvMsgResult = f
-      }
+-- | Get the node server's Node-to-Client version.
+getNtcVersion :: LocalStateQueryExpr block point (QueryInMode mode) r IO (Either QueryError NodeToClientVersion)
+getNtcVersion =  LocalStateQueryExpr . ReaderT $ pure . Right
+
+-- | Lift a query value into a monadic query expression.
+-- Use 'queryExpr' in a do block to construct monadic local state queries.
+queryExpr :: QueryInMode mode a -> LocalStateQueryExpr block point (QueryInMode mode) r IO (Either QueryError a)
+queryExpr q = runExceptT $ do
+  let minNtcVersion = ntcVersionOf q
+  ntcVersion <- ExceptT getNtcVersion
+  if ntcVersion >= minNtcVersion
+    then lift
+      $ LocalStateQueryExpr $ ReaderT $ \_ -> ContT $ \f -> pure $
+        Net.Query.SendMsgQuery q $
+          Net.Query.ClientStQuerying
+          { Net.Query.recvMsgResult = f
+          }
+    else throwE $ QueryErrorUnsupportedVersion minNtcVersion ntcVersion
+
+-- | Lift a query value into a monadic query expression returning Maybe of a result.
+-- This is the same as 'queryExpr' except if the query is not supported by the server, will return Nothing instead
+-- of throwing an error.
+maybeQueryExpr :: QueryInMode mode a -> LocalStateQueryExpr block point (QueryInMode mode) r IO (Either QueryError (Maybe a))
+maybeQueryExpr q = runExceptT $ catchE (Just <$> ExceptT (queryExpr q)) $ \e ->
+  case e of
+    QueryErrorUnsupportedVersion _ _ -> return Nothing
+    _ -> throwE e
```

Comment:
So generally we use: https://hackage.haskell.org/package/transformers-except-0.1.1/docs/Control-Monad-Trans-Except-Extra.html

I think `throwE` is misleading vs `left` as we're not throwing an exception here. What are your thoughts?

Reply:
The main reason I `throwE` here is because I `catchE`, as they are duals of each other.